app/src/forms/form/formMetadata/service.js

```diff
@@ -0,0 +1,104 @@
+const Problem = require('api-problem');
+
+const { v4: uuidv4 } = require('uuid');
+
+const { FormMetadata } = require('../../common/models');
+const { typeUtils } = require('../../common/utils');
+
+const DEFAULT_HEADERNAME = 'X-FORM-METADATA';
+const DEFAULT_ATTRIBUTENAME = 'formMetadata';
+
+const service = {
+  validate: (data) => {
+    if (!data) {
+      throw new Problem(422, `'formMetadata record' cannot be empty.`);
+    }
+  },
+
+  initModel: (formId, data) => {
+    return {
+      id: uuidv4(),
+      formId: formId,
+      headerName: data.headerName ? data.headerName : DEFAULT_HEADERNAME,
+      attributeName: data.attributeName ? data.attributeName : DEFAULT_ATTRIBUTENAME,
+      metadata: data.metadata ? data.metadata : {},
+    };
+  },
+
+  hasMetadata: (formMetadata) => {
+    return formMetadata && formMetadata.metadata && Object.keys(formMetadata.metadata).length;
+  },
+
+  addMetadataToObject: async (formId, data, name, encode) => {
+    if (!formId || !data || !name) return;
+    if (!['attributeName', 'headerName'].includes(name)) return;
+
+    if (data && typeUtils.isObject(data)) {
+      const o = await service.read(formId);
+      if (service.hasMetadata(o)) {
+        let value = o.metadata;
+        if (encode) {
+          let bufferObj = Buffer.from(JSON.stringify(o.metadata), 'utf8');
+          value = bufferObj.toString('base64');
+        }
+        data[o[name]] = value;
+      }
+    }
+  },
+
+  addAttribute: async (formId, obj) => {
+    return await service.addMetadataToObject(formId, obj, 'attributeName');
+  },
+
+  addHeader: async (formId, headers) => {
+    return await service.addMetadataToObject(formId, headers, 'headerName', true);
+  },
+
+  create: async (formId, data, currentUser) => {
+    return service.upsert(formId, data, currentUser);
+  },
+
+  read: async (formId) => {
+    return FormMetadata.query().modify('filterFormId', formId).first();
+  },
+
+  update: async (formId, data, currentUser) => {
+    return service.upsert(formId, data, currentUser);
+  },
+
+  delete: async (formId) => {
+    const existing = await FormMetadata.query().modify('filterFormId', formId).first().throwIfNotFound();
+    await FormMetadata.query().deleteById(existing.id);
+  },
+
+  upsert: async (formId, data, currentUser, transaction) => {
```

Comment:
We don't have Sonar set up yet, but I think this function might fail the cognitive complexity checks in Code Climate.

I think that our usual pattern is that `service.create` does an SQL `insert` and `service.update` does a SQL `modify`. With this function, though, `create` does an `insert` OR a `modify`, and update also does an `insert` OR a `modify`. So POST and PUT are the same thing, yes? While it's not technically wrong for POST to update or PUT to create, it's an inconsistency in the code that would need to be changed later.

Reply:
I will be removing all unused code, which will only leave the upsert function. i can refactor it to reduce complexity.

Reply:
refactored as much as possible. there are no create or update endpoints. should minimize confusion. in order to ensure a single record per form we encapsulate the logic in the upsert method.